pkg/roll/run_update.go

```diff
@@ -87,24 +88,15 @@ func RetryOrQuit(f func() error, quit <-chan struct{}, logger logging.Logger, er
 func (u update) Run(quit <-chan struct{}) (ret bool) {
 	u.logger.NoFields().Debugln("Locking")
 	// TODO: implement API for blocking locks and use that instead of retrying
-	if !RetryOrQuit(u.lockRCs, quit, u.logger, "Could not lock update") {
+	if !RetryOrQuit(
+		func() error { return u.lockRCs(quit) },
+		quit,
+		u.logger,
+		"Could not lock update",
+	) {
 		return
 	}
-
-	// release both locks on termination
-	defer func() {
-		// failure to unlock probably means:
-		// - session was already invalidated, lock was already released when
-		//   that happened
-		// - network error, session will probably time out soon
-		// either way no point retrying
-		if err := u.lock.Unlock(u.lockPath(u.OldRC)); err != nil {
-			u.logger.WithError(err).Warnln("Could not unlock old RC for updates")
-		}
-		if err := u.lock.Unlock(u.lockPath(u.NewRC)); err != nil {
-			u.logger.WithError(err).Warnln("Could not unlock new RC for updates")
-		}
-	}()
+	defer u.unlockRCs(quit)
```

Comment:
The `quit` channel is signaled up in [`Farm.releaseChild()`](https://github.com/square/p2/blob/master/pkg/roll/farm.go#L197).  As to when this "child" update is released, I see three triggers:

1. The session is expiring—all locks have been or will soon be released automatically, and you just want to exit all pending operations as fast as possible.
2. The whole Farm is being asked to quit—the whole server is exiting, and you can treat this like No. 1.
3. The Update object has been removed from Consul—this is the mechanism by which a rolling update can be stopped mid-update, so I'd expect it to happen frequently relative to Nos. 1-2.

The problem is that there isn't a way to distinguish why the update is exiting.  Adding that signal means a lot more complexity here.  Is it worth the extra development time and effort?  I don't know.

I can only point out that `RetryOrQuit()` always performs 1 attempt whether or not `quit` is signaled.